data/tutorials/language/0lg_08_basic_datatypes.md

```diff
@@ -0,0 +1,689 @@
+```ocaml
+# read_line;;
+- : unit -> string = <fun>
+
+# print_endline;;
+- : string -> unit = <fun>
+```
+
+Function `read_line` reads an end-of-line terminated sequence of characters from standard input and returns it as a string. Reading input begins when `()` is passed.
+
+Function `print_endline` prints the string followed by a line ending on standard output. Return of the unit value means the output request has been queued by the operating system.
+
+## User-Defined Types
+
+### Variants
+
+#### Enumerated Data Types
+
+The simplest form of a variant type corresponds to an [enumerated type](https://en.wikipedia.org/wiki/Enumerated_type). It is defined by an explicit list of named values. Defined values are called constructors and must be capitalised.
+
+For example, here is how variant data types could be defined to represent Dungeons & Dragons character classes and alignment.
+```ocaml
+# type character_class =
+    | Barbarian
+    | Bard
+    | Cleric
+    | Druid
+    | Fighter
+    | Monk
+    | Paladin
+    | Ranger
+    | Rogue
+    | Sorcerer
+    | Wizard;;
+type character_class =
+    Barbarian
+  | Bard
+  | Cleric
+  | Druid
+  | Fighter
+  | Monk
+  | Paladin
+  | Ranger
+  | Rogue
+  | Sorcerer
+  | Wizard
+
+# type rectitude = Evil | R_Neutral | Good;;
+type rectitude = Evil | R_Neutral | Good
+
+# type firmness = Chaotic | F_Neutral | Lawful;;
+type firmness = Chaotic | F_Neutral | Lawful
+```
+
+These kinds of variant types can also be used to represent weekdays, cardinal
+directions, or any other fixed-sized set of values that can be given names. A
+total ordering is defined on values following the definition order (e.g., `Druid
+< Ranger`).
+
+Here is how pattern matching can be done on types defined as such.
+```ocaml
+# let rectitude_to_french = function
+    | Evil -> "Mauvais"
+    | R_Neutral -> "Neutre"
+    | Good -> "Bon";;
+val rectitude_to_french : rectitude -> string = <fun>
+```
+
+Note that:
+
+- `unit` is a variant with a unique constructor, which does not carry data: `()`.
+- `bool` is also a variant with two constructors that don't carry data: `true` and `false`.
+
+A pair `(x, y)` has type `a * b`, where `a` is the type of `x` and `b` is the type of `y`. Some may find it intriguing that `a * b` is called a product. Although this is not a complete explanation, here is a remark that may help in understanding: Consider the product type `character_class * character_alignement`. There are 12 classes and 9 alignments. Any pair of values from those types inhabit the product type. Therefore, in the product type, there are 9 × 12 = 108 values, which also is a product.
```

Comment:
```suggestion
A pair `(x, y)` has type `a * b`, where `a` is the type of `x` and `b` is the type of `y`. Some may find it intriguing that `a * b` is called a product. Although this is not a complete explanation, here is a remark that may help in understanding: Consider the product type `character_class * character_alignement`. There are 12 classes and 9 alignments. Any pair of values from those types inhabits the product type. Therefore, in the product type, there are 9 × 12 = 108 values, which is also a product.
```

Reply:
What does "which also is a product" mean? FWIW, this paragraph didn't make things much clearer for me (a beginner). I would just say that it's called a product because it's a Cartesian product, and maybe link to the [wikipedia article](https://en.wikipedia.org/wiki/Cartesian_product).